tesseract_time_parameterization/src/time_optimal_trajectory_generation.cpp

```diff
@@ -0,0 +1,996 @@
+  }
+  else if (max_acceleration_scaling_factor == 0.0)
+  {
+    CONSOLE_BRIDGE_logDebug("A max_acceleration_scaling_factor of 0.0 was specified, defaulting to %f instead.",
+                            acceleration_scaling_factor);
+  }
+  else
+  {
+    CONSOLE_BRIDGE_logWarn("Invalid max_acceleration_scaling_factor %f specified, defaulting to %f instead.",
+                           max_acceleration_scaling_factor,
+                           acceleration_scaling_factor);
+  }
+
+  // Validate limits
+  if (max_velocity.rows() != max_acceleration.rows())
+  {
+    CONSOLE_BRIDGE_logError("Invalid max_velocity or max_acceleration specified. They should be the same length");
+  }
+
+  // Flatten program
+  std::vector<std::reference_wrapper<Instruction>> trajectory = flatten(program, programFlattenMoveInstructionFilter);
+  const Eigen::Index num_joints = max_velocity.rows();
+  const std::size_t num_points = trajectory.size();
+
+  // This lib does not actually work properly when angles wrap around, so we need to unwind the path first
+  //  trajectory.unwind(); /// @todo
+
+  // Have to convert into Eigen data structs and remove repeated points
+  //  (https://github.com/tobiaskunz/trajectories/issues/3)
+  std::list<Eigen::VectorXd> points;
+  for (size_t p = 0; p < num_points; ++p)
+  {
+    auto waypoint = trajectory[p].get().cast_const<MoveInstruction>()->getWaypoint().cast_const<StateWaypoint>();
+    bool diverse_point = (p == 0);
+
+    for (Eigen::Index j = 0; j < num_joints; j++)
+    {
+      if (p > 0 && std::abs(waypoint->position[j] - points.back()[j]) > min_angle_change_)
+        diverse_point = true;
+    }
+
+    if (diverse_point)
+      points.push_back(waypoint->position);
+  }
+
+  // Return trajectory with only the first waypoint if there are not multiple diverse points
+  if (points.size() == 1)
+  {
+    CONSOLE_BRIDGE_logDebug("Trajectory is parameterized with 0.0 dynamics since it only contains a single distinct "
+                            "waypoint.");
+    auto waypoint = trajectory[0].get().cast<MoveInstruction>()->getWaypoint().cast<StateWaypoint>();
+    waypoint->velocity = Eigen::VectorXd::Zero(num_joints);
+    waypoint->acceleration = Eigen::VectorXd::Zero(num_joints);
+    return true;
+  }
+
+  // Now actually call the algorithm
+  totg::Trajectory parameterized(totg::Path(points, path_tolerance_), max_velocity, max_acceleration, 0.001);
+  if (!parameterized.isValid())
+  {
+    CONSOLE_BRIDGE_logError("Unable to parameterize trajectory.");
+    return false;
+  }
+
+  // Compute sample count
+  size_t sample_count = static_cast<std::size_t>(std::ceil(parameterized.getDuration() / resample_dt_));
+
+  // Resample and fill in trajectory
+  auto input_instruction = trajectory[0].get().cast<MoveInstruction>();
+  CompositeInstruction new_program(program);
+  new_program.clear();
+
+  /// @todo Figure out how to maintain the original seed structure of subcomposites
```

Comment:
@Levi-Armstrong I looked into this. I don't think this will be possible without substantial modification. There are several layers we would have to make sure we pass this information through correctly. This also means that applying scaling factors on a PlanInstruction basis won't be possible.